File: src/spladder_core.m
function spladder_core(CFG)
% function spladder_core(CFG)

    %%% add paths
    if isfield(CFG, 'paths'),
        addpath(CFG.paths{:});
    end;

    genes_loaded = 0;

    %%% check if result file exists and start gen graph step if necessary
    if ~exist(CFG.out_fname, 'file'),
        fprintf('Augmenting splice graphs.\n');
        fprintf('=========================\n\n');
        if ~isfield(CFG, 'genes'),
            load(CFG.anno_fname, 'genes');
        else
            genes = CFG.genes ;
        end;

        %%% force closed intervals
        if CFG.is_half_open,
            genes = half_open_to_closed(genes);
            CFG.is_half_open = 0;
        end;

        genes = gen_graphs(genes, CFG) ;

        fprintf('Saving genes to %s\n', CFG.out_fname) ;
        save(CFG.out_fname, 'genes') ;

        genes_loaded = 1;
    else
        fprintf('Augmenting splice graphs already completed.\n') ;
    end ;

    %%% prune splice graph if necessary
    if CFG.do_prune,
        load_fn = CFG.out_fname;
        CFG.out_fname = strrep(CFG.out_fname, '.mat', '_pruned.mat');
        if ~exist(CFG.out_fname, 'file'),
            %%% load genes if not present yet
            if genes_loaded == 0,
                load(load_fn);
            end;

            %%% make splice graphs unique
            genes = uniquify_splicegraph(genes);

            %%% prune graphs
            num_paths_before = count_all_paths(genes);
            genes = prune_graph(genes, bam_fnames);
            num_paths_after = count_all_paths(genes);

            %%% save pruned genes
            fprintf('saving genes to %s\n', CFG.out_fname) ;
            save(CFG.out_fname, 'genes') ;

            genes_loaded = 1;
        else
            fprintf('Pruning of splice graphs already done\n');
        end;
    else
        fprintf('No pruning requested!\n');
    end;

    %%% generate isoforms if necessary
    if CFG.do_gen_isoforms,
        load_fn = CFG.out_fname;
        CFG.out_fame = strrep(CFG.out_fname, '.mat', '_with_isoforms.mat');
        if ~exist(CFG.out_fname, 'file'),
            %%% load genes if not present yet
            if genes_loaded == 0,
                load(load_fn);
            end;

            %%% generate isoforms
            fprintf('Generating all isoforms\n') ;
            genes = generate_isoforms(genes, PAR.conf);

            %%% re-constitute splicing graph
            fprintf('\tRe-constituting simplified splice graph from generated transcripts\n') ;
            genes_unsimplified = genes ;
            genes = splice_graph(genes, conf.do_infer_splice_graph);

            %%% save splicing graph with isoforms
            fprintf('\tSaving genes to %s\n', CFG.out_fname) ;
            save(CFG.out_fname, 'genes', 'genes_unsimplified') ;
        else
            fprintf('Generating all isoforms already done\n');
        end;
    else
        fprintf('Generating all isoforms not requested\n');
    end;
    fprintf('\n');

File: python/spladder/core/spladdercore.py
import sys
import os
import scipy as sp
import cPickle

from gen_graphs import gen_graphs 

def spladder_core(CFG):

    genes_loaded = False

    ### check if result file exists and start gen graph step if necessary
    if not os.path.exists(CFG['out_fname']):
        print >> sys.stdout, 'Augmenting splice graphs.'
        print >> sys.stdout, '========================='
        if not 'genes' in CFG:
            genes = cPickle.load(open(CFG['anno_fname'], 'r'))
        else:
            genes = CFG['genes']

        genes = gen_graphs(genes, CFG)

        print 'Saving genes to %s' % (CFG['out_fname'])
        cPickle.dump(genes, open(CFG['out_fname'], 'w'), -1)

        genes_loaded = True
    else:
        print 'Augmenting splice graphs already completed.'

    ### prune splice graph if necessary
    if CFG['do_prune']:
        load_fn = CFG['out_fname']
        CFG['out_fname'] = re.sub(r'.pickle$', '_pruned.pickle', CFG['out_fname'])
        if not os.path.exists(CFG['out_fname']):
            ### load genes if not present yet
            if not genes_loaded:
                genes = cPickle.load(open(load_fn), 'r')

            ### make splice graphs unique
            genes = uniquify_splicegraph(genes)

            ### prune graphs
            num_paths_before = count_all_paths(genes)
            genes = prune_graph(genes, bam_fnames)
            num_paths_after = count_all_paths(genes)

            ### save pruned genes
            print 'saving genes to %s' % (CFG['out_fname'])
            cPickle.dump(genes, open(CFG['out_fname'], 'w'), -1)

            genes_loaded = True;
        else:
            print 'Pruning of splice graphs already done'
    else:
        print 'No pruning requested!'

    ### generate isoforms if necessary
    if CFG['do_gen_isoforms']:
        load_fn = CFG['out_fname']
        CFG['out_fame'] = re.sub(r'.pickle$', '_with_isoforms.pickle', CFG['out_fname']);
        if not os.path.exists(CFG['out_fname']):
            ### load genes if not present yet
            if not genes_loaded:
                genes = cPickle.load(open(load_fn))

            ### generate isoforms
            print 'Generating all isoforms'
            genes = generate_isoforms(genes, PAR['conf'])

            ### re-constitute splicing graph
            print '\tRe-constituting simplified splice graph from generated transcripts'
            genes_unsimplified = genes
            genes = splice_graph(genes, CFG['do_infer_splice_graph'])

            ### save splicing graph with isoforms
            print '\tSaving genes to %s' % CFG['out_fname']
            cPickle.dump((genes, genes_unsimplified), open(CFG['out_fname'], 'w'), -1)
        else:
            print 'Generating all isoforms already done'
    else:
        print 'Generating all isoforms not requested'
